hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/CarbonitedMetadataThawingActor.scala

```diff
@@ -2,96 +2,148 @@ package cromwell.services.metadata.hybridcarbonite
 
 import akka.actor.{ActorRef, LoggingFSM, Props, Status}
 import akka.pattern.pipe
+import cats.data.NonEmptyList
+import cromwell.core.WorkflowId
 import cromwell.core.io.{AsyncIo, DefaultIoCommandBuilder}
-import cromwell.core.{RootWorkflowId, WorkflowId}
-import cromwell.services.metadata.MetadataService.{GetRootAndSubworkflowLabels, RootAndSubworkflowLabelsLookupFailed, RootAndSubworkflowLabelsLookupResponse}
+import cromwell.services.metadata.MetadataService._
 import cromwell.services.metadata.hybridcarbonite.CarbonitedMetadataThawingActor._
+import cromwell.services.metadata.{MetadataQuery, MetadataQueryJobKey}
+import cromwell.services.{BuiltMetadataResponse, FailedMetadataResponse}
 import cromwell.util.JsonEditor
 import io.circe.parser._
 import io.circe.{Json, Printer}
+import spray.json._
 
 import scala.concurrent.{ExecutionContext, Future}
 
-final class CarbonitedMetadataThawingActor(carboniterConfig: HybridCarboniteConfig, serviceRegistry: ActorRef, ioActor: ActorRef) extends LoggingFSM[CarbonitedMetadataThawingState, CarbonitedMetadataThawingData] {
+final class CarbonitedMetadataThawingActor(carboniterConfig: HybridCarboniteConfig, serviceRegistry: ActorRef, ioActor: ActorRef) extends LoggingFSM[ThawingState, ThawingData] {
 
   implicit val ec: ExecutionContext = context.dispatcher
 
   val asyncIo = new AsyncIo(ioActor, DefaultIoCommandBuilder)
 
-  startWith(CarbonitedMetadataThawingActorPendingState, CarbonitedMetadataThawingActorPendingData)
+  startWith(PendingState, PendingData)
 
-  when(CarbonitedMetadataThawingActorPendingState) {
-    case Event(ThawCarbonitedMetadata(rootWorkflowId), CarbonitedMetadataThawingActorPendingData) =>
-
-      val response = for {
-        rawResponseFromIoActor <- asyncIo.contentAsStringAsync(carboniterConfig.makePath(rootWorkflowId), maxBytes = Option(30 * 1024 * 1024), failOnOverflow = true)
+  when(PendingState) {
+    case Event(action: WorkflowMetadataReadAction, PendingData) =>
+      val gcsResponse = for {
+        rawResponseFromIoActor <- asyncIo.contentAsStringAsync(carboniterConfig.makePath(action.workflowId), maxBytes = Option(30 * 1024 * 1024), failOnOverflow = true)
         parsedResponse <- Future.fromTry(parse(rawResponseFromIoActor).toTry.map(GcsMetadataResponse.apply))
       } yield parsedResponse
 
-      response pipeTo self
-      serviceRegistry ! GetRootAndSubworkflowLabels(rootWorkflowId)
+      gcsResponse pipeTo self
+
+      val targetState = if (action.requiresLabels) {
+        serviceRegistry ! GetRootAndSubworkflowLabels(action.workflowId)
```

Comment:
⚠️ TOL: side effect during `val` assignment.

I know your way is terser than my alternative, but it reduces the risk of someone (ie me) missing the side effect because it starts with `val ... = ...`
```
if (action.requiresLabels) { serviceRegistry ! GetRootAndSubworkflowLabels(action.workflowId) }
val targetState = if (action.requiresLabels) NeedsGcsAndLabelsState else NeedsOnlyGcsState
```

Reply:
Yeah I didn't really like either of the options; you seem to have a stronger preference than I do so let's go with that 🙂